File: 0000xx/000008/Solution.java
import java.lang.AssertionError;
import java.util.*;
import java.util.regex.*;

// can be easy done with long64 type
// but done by int32 here only

class Solution {

    static final int MIN = Integer.MIN_VALUE;
    static final int MAX = Integer.MAX_VALUE;
    
    static final int BIG10 = 214748364;

    public int myAtoi(String str) {
        char[] cAry = str.toCharArray();
    
        int i=0;
        for(;i<cAry.length;++i){
            if(cAry[i]!=' ')break;
        }
        
        if(i==cAry.length)return 0;
        
        boolean positive = true;
        if(cAry[i]=='-'){positive=false;++i;}
        else if(cAry[i]=='+'){++i;}

        if(i==cAry.length)return 0;
        
        int retAbs = 0;
        
        for(;i<cAry.length;++i){
            char c = cAry[i];
            if(c<'0')break;
            if(c>'9')break;
            
            if(retAbs>BIG10){
                return positive?MAX:MIN;
            }else if(retAbs==BIG10){
                if(( positive)&&(c>='7'))return MAX;
                if((!positive)&&(c>='8'))return MIN;
            }
            
            retAbs*=10;
            retAbs+=(c-'0');
        }
        
        return positive?(retAbs):(-retAbs);
    }
}

File: 0000xx/000008/kotlin/Solution.kt
class Solution {
    fun myAtoi(s: String): Int {
        val BIG_POS:Long = Integer.MAX_VALUE.toLong()
        val SMALL_POS:Long = Integer.MIN_VALUE.toLong()*(-1)
        val cAry = s.toCharArray()
        
        var i=0

        // step 1
        while(i<cAry.size){
            if(cAry[i]!=' '){break}
            ++i
        }
        
        // step 2
        var sign:Char='+'
        if(i<cAry.size){
            if((cAry[i]=='+')||(cAry[i]=='-')){
                sign=cAry[i]
                ++i
            }
        }
        
        // step 3
        var ret:Long=0
        while(i<cAry.size){
            val c=cAry[i]
            if(c<'0'){break}
            if(c>'9'){break}
            ret*=10
            ret+=(c-'0')
            if((sign=='+')&&(ret>BIG_POS)){break}
            if((sign=='-')&&(ret>SMALL_POS)){break}
            ++i
        }
        
        // ret
        if(sign=='-'){
            ret*=-1
        }
        ret = kotlin.math.min(ret,Integer.MAX_VALUE.toLong())
        ret = kotlin.math.max(ret,Integer.MIN_VALUE.toLong())
        
        return ret.toInt()
    }
}
